mirrord/intproxy/src/proxies/incoming/interceptor.rs

```diff
@@ -779,4 +836,48 @@ mod test {
             }
         }
     }
+
+    /// Checks that [`hyper`] and [`h2`] crate versions are in sync with each other.
+    ///
+    /// As we use `source.downcast_ref::<h2::Error>` to drill down on [`h2`] errors from
+    /// [`hyper`], we need these two crates to stay in sync, otherwise we could always
+    /// fail some of our checks that rely on this `downcast` working.
+    ///
+    /// Even though we're using [`h2::Error::is_reset`] in intproxy, this test can be
+    /// for any error, and thus here we do it for [`h2::Error::is_go_away`] which is
+    /// easier to trigger.
+    #[tokio::test]
+    async fn hyper_and_h2_versions_in_sync() {
+        let notify = Arc::new(Notify::new());
+        let wait_notify = notify.clone();
+
+        tokio::spawn(async move {
+            let listener = TcpListener::bind("127.0.0.1:6666").await.unwrap();
+
+            notify.notify_waiters();
+            let (io, _) = listener.accept().await.unwrap();
+
+            if let Err(fail) = hyper::server::conn::http2::Builder::new(TokioExecutor::default())
+                .serve_connection(
+                    TokioIo::new(io),
+                    service_fn(|_| async move {
+                        Ok::<_, Infallible>(Response::new(Full::new(Bytes::from("Heresy!"))))
+                    }),
+                )
+                .await
+            {
+                assert!(fail
+                    .source()
+                    .and_then(|source| source.downcast_ref::<h2::Error>())
+                    .is_some_and(h2::Error::is_go_away));
+            } else {
+                panic!("The request is supposed to fail with `GO_AWAY`, something is wrong if it didn't!");
```

Comment:
This panic could mention that deps versions must be in sync - will be easier to spot in the CI